src/lib.rs

```diff
@@ -127,6 +129,27 @@ impl Multiaddr {
         self
     }
 
+    /// Parses a [`Multiaddr`] from the given string, ensuring that it ends with [`Protocol::P2p`] of the provided peer.
+    ///
+    /// If the address does not end with `/p2p`, it is simply appended.
+    /// If the address ends with a _different_ peer ID, parsing will fail.
+    pub fn from_str_and_peer(addr: &str, peer: PeerId) -> Result<Self> {
+        let addr = addr.parse::<Self>()?;
+        addr.with_p2p(peer).map_err(|_| Error::InvalidMultiaddr)
```

Comment:
Interesting! I didn't expect that we can compose this together. That makes me think we don't need this function actually?

Would reduce our public API surface and as we can see, it is really easy to write and could even be chained!

Reply:
I agree. I'm removing it.